pkg/controller/eventinstance/truncate.go

```diff
@@ -0,0 +1,163 @@
+package eventinstance
+
+import (
+	"context"
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/acorn-io/baaah/pkg/router"
+	internalv1 "github.com/acorn-io/runtime/pkg/apis/internal.acorn.io/v1"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func Truncate(ctx context.Context, c kclient.Client, capacity int) router.HandlerFunc {
+	t := &truncator{
+		client:   c,
+		capacity: capacity,
+		tracking: make(map[kclient.ObjectKey]struct{}),
+		trigger:  make(chan struct{}, 1),
+		mu:       new(sync.Mutex),
+	}
+	var once sync.Once
+
+	return func(req router.Request, _ router.Response) error {
+		once.Do(func() {
+			// Lazy start the truncator.
+			// This ensures it's started after the router cache has synced
+			go func() {
+				if err := t.start(ctx); err != nil {
+					logrus.WithError(err).Warn("Exiting eventinstance truncator")
+				}
+			}()
+		})
+
+		eventKey := router.Key(req.Namespace, req.Name)
+		if req.Object != nil {
+			// Add or Update, add to the running count of events
+			t.track(eventKey)
+		} else {
+			// Delete, remove from the running count of events
+			t.forget(eventKey)
+		}
+
+		return nil
+	}
+}
+
+type truncator struct {
+	client   kclient.Client
+	capacity int
+	tracking map[kclient.ObjectKey]struct{}
+	trigger  chan struct{}
+	mu       *sync.Mutex
+}
+
+func (t *truncator) start(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.trigger:
+		}
+
+		if err := t.truncate(ctx); err != nil {
+			logrus.WithError(err).Warn("Failed to truncate eventinstances, retrying in 15 seconds")
+			go func() {
+				time.Sleep(15 * time.Second)
+				t.kick()
+			}()
+			continue
+		}
+
+		// Give the router a chance to process more events
+		time.Sleep(1 * time.Minute)
+	}
+}
+
+func (t *truncator) kick() {
+	select {
+	case t.trigger <- struct{}{}:
+	default:
+	}
+}
+
+func (t *truncator) track(eventKey kclient.ObjectKey) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.tracking[eventKey] = struct{}{}
+
+	if len(t.tracking) > t.capacity {
+		t.kick()
+	}
+}
+
+func (t *truncator) forget(eventKey kclient.ObjectKey) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	delete(t.tracking, eventKey)
+}
+
+func (t *truncator) truncate(ctx context.Context) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if len(t.tracking) <= t.capacity {
+		// Nothing to do
+		return nil
+	}
+
+	var events internalv1.EventInstanceList
+	if err := t.client.List(ctx, &events); err != nil {
+		return err
+	}
```

Comment:
alternatively, we keep a sorted slice with ordered inserts on add, but I'm not sure that would be consistent with the state of our cache since we're consuming events from the queue concurrently